external/kafka/src/main/scala/org/apache/spark/rdd/kafka/KafkaCluster.scala

```diff
@@ -0,0 +1,313 @@
+      val respMap = resp.requestInfo
+      val needed = topicAndPartitions.diff(result.keys.toSet)
+      needed.foreach { tp =>
+        respMap.get(tp).foreach { offsetMeta =>
+          if (offsetMeta.error == ErrorMapping.NoError) {
+            result += tp -> offsetMeta
+          } else {
+            errs.append(ErrorMapping.exceptionFor(offsetMeta.error))
+          }
+        }
+      }
+      if (result.keys.size == topicAndPartitions.size) {
+        return Right(result)
+      }
+    }
+    val missing = topicAndPartitions.diff(result.keys.toSet)
+    errs.append(new Exception(s"Couldn't find consumer offsets for ${missing}"))
+    Left(errs)
+  }
+
+  def setConsumerOffsets(
+    groupId: String,
+    offsets: Map[TopicAndPartition, Long]
+  ): Either[Err, Map[TopicAndPartition, Short]] = {
+    setConsumerOffsetMetadata(groupId, offsets.map { kv =>
+      kv._1 -> OffsetMetadataAndError(kv._2)
+    })
+  }
+
+  def setConsumerOffsetMetadata(
+    groupId: String,
+    metadata: Map[TopicAndPartition, OffsetMetadataAndError]
+  ): Either[Err, Map[TopicAndPartition, Short]] = {
+    var result = Map[TopicAndPartition, Short]()
+    val req = OffsetCommitRequest(groupId, metadata)
+    val errs = new Err
+    val topicAndPartitions = metadata.keys.toSet
+    withBrokers(seedBrokers, errs) { consumer =>
+      val resp = consumer.commitOffsets(req)
+      val respMap = resp.requestInfo
+      val needed = topicAndPartitions.diff(result.keys.toSet)
+      needed.foreach { tp =>
+        respMap.get(tp).foreach { err =>
+          if (err == ErrorMapping.NoError) {
+            result += tp -> err
+          } else {
+            errs.append(ErrorMapping.exceptionFor(err))
+          }
+        }
+      }
+      if (result.keys.size == topicAndPartitions.size) {
+        return Right(result)
+      }
+    }
+    val missing = topicAndPartitions.diff(result.keys.toSet)
+    errs.append(new Exception(s"Couldn't set offsets for ${missing}"))
+    Left(errs)
+  }
+
+  private def withBrokers(brokers: Iterable[(String, Int)], errs: Err)
+    (fn: SimpleConsumer => Any): Unit = {
+    brokers.foreach { hp =>
+      var consumer: SimpleConsumer = null
+      try {
+        consumer = connect(hp)
+        fn(consumer)
+      } catch {
+        case NonFatal(e) =>
+          errs.append(e)
+      } finally {
+        if (consumer != null) consumer.close()
+      }
+    }
+  }
+}
+
+object KafkaCluster {
+  type Err = ArrayBuffer[Throwable]
+
+  /** Make a consumer config without requiring group.id or zookeeper.connect,
+    * since communicating with brokers also needs common settings such as timeout
+    */
+  def consumerConfig(kafkaParams: Map[String, String]): ConsumerConfig = {
+    val props = new Properties()
+    kafkaParams.foreach(param => props.put(param._1, param._2))
+    Seq("zookeeper.connect", "group.id").foreach { s =>
+      if (!props.contains(s)) {
+        props.setProperty(s, "")
+      }
+    }
```

Comment:
I wouldn't use the collect (even though it allows you to do the partial function) because it is actually obfuscating the intention here, because it is perceived as a filter. foreach makes more sense.